components/style/stylesheet_set.rs

```diff
@@ -56,26 +79,34 @@ where
     iter: slice::IterMut<'a, StylesheetSetEntry<S>>,
     guard: &'a SharedRwLockReadGuard<'b>,
     origins_dirty: OriginSet,
+    origin_data_validity: PerOrigin<OriginValidity>,
     author_style_disabled: bool,
     had_invalidations: bool,
 }
 
 /// The type of rebuild that we need to do for a given stylesheet.
 pub enum SheetRebuildKind {
-    /// For now we only support full rebuilds, in the future we'll implement
-    /// partial rebuilds.
+    /// A full rebuild, of both cascade data and invalidation data.
     Full,
+    /// A partial rebuild, of only the cascade data.
+    CascadeOnly,
+}
+
+impl SheetRebuildKind {
+    /// Whether the stylesheet invalidation data should be rebuilt.
+    pub fn rebuild_invalidation(&self) -> bool {
```

Comment:
Nit: function names sounds a bit like it's going to do the rebuilding. Maybe "must_rebuild_invalidation" or something?